File: services/core-api/src/domain/read-model/execution-read-model.ts
/**
 * Execution Read Model（UI向け正規形）
 * data-integration-contract §2.1 に準拠。
 * UI はこの型のフィールドのみを前提に描画する。
 */

export type ExecutionReadModelStatus = "ACTIVE" | "COMPLETED" | "FAILED" | "CANCELED";

export type ExecutionReadModelNodeStatus =
  | "IDLE"
  | "READY"
  | "RUNNING"
  | "WAITING"
  | "SUCCEEDED"
  | "FAILED"
  | "CANCELED";

export type ExecutionReadModelNode = {
  nodeId: string;
  nodeType: string;
  status: ExecutionReadModelNodeStatus;
  attempt: number;
  workerId: string | null;
  waitKey: string | null;
  canceledByExecution: boolean;
  /** 契約 §2.1 の必須外。後方互換のため API は返却してよい。 */
  error?: { message?: string | null } | null;
};

export type ExecutionReadModel = {
  executionId: string;
  graphId: string;
  status: ExecutionReadModelStatus;
  cancelRequestedAt: string | null;
  canceledAt: string | null;
  failedAt: string | null;
  completedAt: string | null;
  nodes: ExecutionReadModelNode[];
};

/**
 * ExecutionState（内部）を ExecutionReadModel（UI向け）に変換する。
 * 契約に含まないフィールド（version, node.output, node.error）は含めない。
 */
export function toExecutionReadModel(state: {
  executionId: string;
  graphId: string;
  status: string;
  cancelRequestedAt?: string | null;
  canceledAt?: string | null;
  failedAt?: string | null;
  completedAt?: string | null;
  nodes: Record<
    string,
    {
      nodeId: string;
      nodeType: string;
      status: string;
      attempt: number;
      workerId?: string | null;
      waitKey?: string | null;
      canceledByExecution?: boolean;
      error?: unknown;
    }
  >;
}): ExecutionReadModel {
  return {
    executionId: state.executionId,
    graphId: state.graphId,
    status: state.status as ExecutionReadModelStatus,
    cancelRequestedAt: state.cancelRequestedAt ?? null,
    canceledAt: state.canceledAt ?? null,
    failedAt: state.failedAt ?? null,
    completedAt: state.completedAt ?? null,
    nodes: Object.values(state.nodes).map((n) => {
      const err = n.error;
      const errorShape =
        err != null && typeof err === "object" && "message" in err
          ? { message: typeof (err as { message?: unknown }).message === "string" ? (err as { message: string }).message : null }
          : null;
      return {
        nodeId: n.nodeId,
        nodeType: n.nodeType,
        status: n.status as ExecutionReadModelNodeStatus,
        attempt: n.attempt,
        workerId: n.workerId ?? null,
        waitKey: n.waitKey ?? null,
        canceledByExecution: n.canceledByExecution ?? false,
        ...(errorShape === null ? {} : { error: errorShape })
      };
    })
  };
}

File: api/Statevia.Core.Api/Contracts/ExecutionReadModel.cs
using System;
using System.Collections.Generic;

namespace Statevia.Core.Api.Contracts;

/// <summary>Execution Read Model（UI 向け正規形）。</summary>
public sealed class ExecutionReadModel
{
    /// <summary>実行 ID（表示用）。</summary>
    public string ExecutionId { get; init; } = string.Empty;

    /// <summary>グラフ（定義）ID。</summary>
    public string GraphId { get; init; } = string.Empty;

    /// <summary>実行全体の状態。</summary>
    public string Status { get; init; } = string.Empty;

    /// <summary>キャンセル要求日時（任意）。</summary>
    public DateTimeOffset? CancelRequestedAt { get; init; }

    /// <summary>キャンセル完了日時（任意）。</summary>
    public DateTimeOffset? CanceledAt { get; init; }

    /// <summary>失敗日時（任意）。</summary>
    public DateTimeOffset? FailedAt { get; init; }

    /// <summary>完了日時（任意）。</summary>
    public DateTimeOffset? CompletedAt { get; init; }

    // TODO: nodes の詳細は ExecutionGraph の仕様に合わせて埋める（現状は空コレクションを返す実装になる想定）。
    /// <summary>実行ノードの列。</summary>
    public IReadOnlyList<ExecutionNodeReadModel> Nodes { get; init; } =
        Array.Empty<ExecutionNodeReadModel>();
}

/// <summary>実行読み取りモデル上の 1 ノード。</summary>
public sealed class ExecutionNodeReadModel
{
    /// <summary>ExecutionGraph のノード識別子（試行単位）。</summary>
    public string ExecutionNodeId { get; init; } = string.Empty;

    /// <summary>ノード種別。</summary>
    public string NodeType { get; init; } = string.Empty;

    /// <summary>ノードの状態。</summary>
    public string Status { get; init; } = string.Empty;

    /// <summary>試行番号。</summary>
    public int Attempt { get; init; }

    /// <summary>ワーカー ID（任意）。</summary>
    public string? WorkerId { get; init; }

    /// <summary>Wait キー（任意）。</summary>
    public string? WaitKey { get; init; }

    // TODO: canceledByExecution は ExecutionGraph 側に情報を持たせた上でマッピングする。
    /// <summary>実行キャンセルにより打ち切られたか。</summary>
    public bool CanceledByExecution { get; init; }
}
